[WIP] matrices: trace of MatAdd

lokeshh opened the pull request:
Earlier trace(MatAdd(X, Y)) was failing with type error
'mix of Matrix and Scalar symbols'. This was because
trace(MatAdd(X, Y)) evaluated to MatAdd(trace(X), trace(Y))
and since validate function would reject trace(X) as Matrix
object it was throwing an exception.

It has been fixed by evaluating trace(MatAdd(X, Y)) to
trace(Z) where Z is what we have after evaluating
MatAdd(X, Y).

Fixes issue #9028

cbm755 committed:
Core: inequalities with NaN raise TypeError

NaN is indeterminant; for example the result of `0/0` could be
many values in the sense of limits (`5*x/x`, `x/x`, as `x` to zero).
By this interpretation, `3 < NaN` is not an valid comparison and
we raise the error.

Note this is different from the usual IEEE 754 Floating Point
system which regards all inequalities with NaN as false, using
the logic "3 is not determinably less than NaN".

Some tests in core/tests/test_numbers also updated.

This is discussed in issue #7773 and in pr #7776.

cbm755 committed:
meijerint: workaround NaN's in _check_antecedents

Inequality checks with NaN's used to be false.  Now they raise
TypeError.  This causes failures in meijerint.py's
`_check_antecedents`.  I've kludged it to get the previous behaviour
but its possible those NaN's are a symptom of a real issue.  Left
some comments and FIXME's in there for this.  Also labelled several
of the tests that hit the issue.

cbm755 committed:
NaN: docstring update: doesn't exactly "represent" floating point nan

With changes to inequality behaviour, NaN behaves differently than
floating point nan.  We say "loosely related" instead.

cbm755 committed:
Infinity/NegativeInfinity: simplify relational code

If we call `Expr.__lt__` then we do not need as much error checking
here.  In particular, can drop the "compare to nonreal" error.

Rebasing

dqnykamp committed:
In parse_expr with implicit multiplication, check if parentheses match.

smichr committed:
8907: improve comp for testing for (in)significant difference

A test as simple as Mod(1.3, 1) == .3 fails in SymPy because the
underlying Floats that end up being compared are not exactly the
same. To fascillitate such comparisons, changes were made to the
comp function. To compare such numbers the comp function can be
used as comp(Mod(1.3, 1), .3) and the result will be True.

leosartaj committed:
Fix #8984 - Use \left \| instead of \lvert in latex(Abs)

cbm755 committed:
MatPow: add doit

cbm755 committed:
MatPow: add doit

cbm755 committed:
MatAdd: doit() drills down in args

Previously, doit() would only make superficial changes, even with
`deep=True` specified.  Now it does.  Add tests.
We default to `deep=True` in contrast with MatMul, will investigate
that discrepancy later, with a view to changing MatMul...

cbm755 committed:
MatPow: doit can deal with some MatrixSymbols too

cbm755 committed:
MatPow: doit can deal with some MatrixSymbols too

Trace: Improved expression evaluation involving Trace

Previouslly, expression involding trace weren't giving the right results.
For example, trace(Y) + trace(-Y) wasn't giving the desired result
that is zero because trace(-Y) was equivalent to
"Mul(Integer(-1), Trace(MatMul(MatrixSymbol('Y', Integer(1), Integer(1)))))"
The presence of MatMul was not required and was the reason things weren't going
as they should.

It has been rectified by modifying the MatMul._eval_trace function.
The code is much cleaner, easy to understand and giving correct results now.
Now trace(-Y) is equivalent to
"Mul(Integer(-1), Trace(MatrixSymbol('Y', Integer(1), Integer(1))))"
which is both correct and upon addition with trace(Y) is quivalent to zero
like it should be.

8907: improve comp for testing for (in)significant difference

A test as simple as Mod(1.3, 1) == .3 fails in SymPy because the
underlying Floats that end up being compared are not exactly the
same. To fascillitate such comparisons, changes were made to the
comp function. To compare such numbers the comp function can be
used as comp(Mod(1.3, 1), .3) and the result will be True.

Fixing conventions of LaTeX printing for rf/ff

Infinity/NegativeInfinity: simplify relational code

If we call `Expr.__lt__` then we do not need as much error checking
here.  In particular, can drop the "compare to nonreal" error.

NaN: docstring update: doesn't exactly "represent" floating point nan

With changes to inequality behaviour, NaN behaves differently than
floating point nan.  We say "loosely related" instead.

watch for inequality TypeError in valid

fix #8984 - use \left | instead of \lvert in latex(Abs)

raise TypeError for inequalities involving zoo

and Except TypeError: invalid comparison of zoo in _simplifyconds

minor refactoring in meijerint

MatMul, Trace: various tests and XFAIL tests

cbm755 committed:
MatPow: doit tests

cbm755 committed:
MatPow: as_explicit rewrite in terms of doit

cbm755 committed:
MatPow: as_explicit additional tests for MatrixSymbols

cbm755 committed:
MatMul: make doit deep=True by default

I don't see why MatMul gets special treatment.  Now the behaviour
is similar to other classes; `A.doit()` now applies `doit()` to
its arguments.  You can still pass `deep=False` if you don't want
it to drill down.

Closes #8138.  Not exactly fixes, but makes it unnecessary.

cbm755 committed:
MatMul: tests cleanup, remove unnecessary variables, add whitespace

Commit:
Fixing conventions of LaTeX printing for rf/ff

aktech committed:
Extend Catalan number to negative numbers

aktech committed:
Add tests for Extend Catalan number to negative numbers

cbm755 committed:
MatrixExpr: Trace: make doit deep=True by default

Also addes a test.

toolforger committed:
Removed C from integrals

.

toolforger committed:
Removed C from integrals

cbm755 committed:
MatrixExpr: Trace: add XFAIL test for trace of MatAdd

toolforger committed:
Remove C from printing

toolforger committed:
Removed C from series

toolforger committed:
Removed C from simplify

toolforger committed:
Removed C from solvers where the imports would not create cycles

cswiercz committed:
rootof: fix numerical evaluation of RootOf

RootOf._eval_evalf() hangs in the case when the computed root lies on
the boundary of the interval containing it. This occurs even in the
following simple example.

In [1]: r = RootOf(x**2 + 1, 0, radicals=False)
In [2]: r.evalf()   # hangs indefinitely

The solution is to allow for such boundary interval values.
Fixes #9019. The above example is included as a test.

toolforger committed:
Remove C by using local imports.

evalf table creation runs just once, so it can locally import freely without
incurring measurable overhead.
The other local imports replace another local import, so the change cannot
cause any performance loss.

smichr committed:
add Lokesh Sharma to AUTHORS/aboutus/.mailmap. Welcome to SymPy!

smichr committed:
add Sartaj Singh to AUTHORS/aboutus. Welcome to SymPy!

lokeshh committed:
matrices: trace of MatAdd

Earlier trace(MatAdd(X, Y)) was failing with type error
'mix of Matrix and Scalar symbols'. This was because
trace(MatAdd(X, Y)) evaluated to MatAdd(trace(X), trace(Y))
and since validate function would reject trace(X) as Matrix
object it was throughing an exception.

It has been fixed by evaluating trace(MatAdd(X, Y)) to
trace(Z) where Z is what we have after evaluating
MatAdd(X, Y).

Fixes #9028

cbm755 commented:
Awesome, thanks for looking into this.

Can you hold off until after #9009?  It has an XFAIL test for this, which this PR should be able to clean up too.

I was going to change it to `return Add(*[Trace(arg).doit() for arg in self.args]).doit()` but your approach seems simpler!  But should we worry about the following?

````
X = Matrix([[1, 2], [3, 4]]);
Y = MatrixSymbol('Y', 2, 2);
q = MatAdd(X, Y, 2*X)
Trace(q).doit()
Out[45]: 
Trace(Matrix([
[3,  6],
[9, 12]]) + Y)
````
In my proposed PR that goes to `15 + Trace(y)`.  Is that better?  Or is `Add` dangerous here somehow?

cbm755 commented:
I guess I don't see the idea of the trace distributing across the addition in your patch.  OTOH, it does fix the issue I was having...  I'm not sure.  What do you think?

lokeshh commented:
After seeing your way, I would prefer your way of doing things but I think there is still some work to be done. 

For instance,

```python
    X = Matrix([1])
    Y = MatrixSymbol('Y', 1, 1)
    >>> trace(X + Y - Y)
    1                   # This looks good :)
    >>> trace(MatAdd(X, Y, -Y))
    1 - Trace(Y) + Trace(Y)                    # This doesn't look good :(
```

jksuom committed:
issue #8972: patch for ode classification failure in pypy

Make the order of functions independent of dictionary order.

jksuom committed:
issue #8972: patch for ode classification failure in pypy

Rebasing

smichr committed:
add Chris Swierczewski to AUTHORS/aboutus. Welcome to SymPy!

lokeshh commented:
It seems the problem is with `trace` not `Add`. 

Take a look at this:
```python
>>> Y = MatrixSymbol('Y', 1, 1)
>>> trace(-Y)
-Trace(Y)
>>> -trace(Y)
-Trace(Y)
>>> trace(-Y) == -trace(Y)
False

cbm755 commented:
Well perhaps we don't need to fix everything all at once ;-)  I like the idea of adding an XFAIL test at least for your first `X + Y - Y` example.  You could file an issue, then put the issue number as a comment in the test.  #9009 is merged so this PR could also remove the XFAIL decorator from my test in tests/test_trace

Do you want to do those things?  Let me know if you need help.

Also: take a look at the `srepr` for your second example.  Does this still happen after #9009?  I think doit(deep) would take the MatrixSymbol out of the MatMul, but have not tried yet.

cbm755 commented:
Also, #9009 had fixes for MatAdd so perhaps things look better now :)

cbm755 commented:
By the way: add "Fixes #9028." in the body of one of your commit messages and when this is merged, it will automatically close the bug.

smichr committed:
remove bare except

smichr committed:
remove bare except

lokeshh committed:
Switch to Add(trace) instead of trace(MatAdd)

Inorder for trace to be distributed all over the expression
to have flexibility in evaluation, distributing MatAdd over
trace wasn't helpful.

lokeshh committed:
# This is a combination of 5 commits.
# The first commit's message is:
# This is a combination of 2 commits.
# The first commit's message is:

# This is a combination of 16 commits.
# The first commit's message is:
Switch to Add(trace) instead of trace(MatAdd)

Inorder for trace to be distributed all over the expression
to have flexibility in evaluation, distributing MatAdd over
trace wasn't helpful.

# This is the 2nd commit message:

Trace: Improved expression evaluation involving Trace

Previouslly, expression involding trace weren't giving the right results.
For example, trace(Y) + trace(-Y) wasn't giving the desired result
that is zero because trace(-Y) was equivalent to
"Mul(Integer(-1), Trace(MatMul(MatrixSymbol('Y', Integer(1), Integer(1)))))"
The presence of MatMul was not required and was the reason things weren't going
as they should.

It has been rectified by modifying the MatMul._eval_trace function.
The code is much cleaner, easy to understand and giving correct results now.
Now trace(-Y) is equivalent to
"Mul(Integer(-1), Trace(MatrixSymbol('Y', Integer(1), Integer(1))))"
which is both correct and upon addition with trace(Y) is quivalent to zero
like it should be.

# This is the 3rd commit message:

Update Matrice Test

# This is the 4th commit message:

Fix the trace test

Forgot to import the ImmutableMatrix module before using it
was the cause tests were passing.
It's been fixed.
Futher Trace doit test has been introduced.

# This is the 5th commit message:

8907: improve comp for testing for (in)significant difference

A test as simple as Mod(1.3, 1) == .3 fails in SymPy because the
underlying Floats that end up being compared are not exactly the
same. To fascillitate such comparisons, changes were made to the
comp function. To compare such numbers the comp function can be
used as comp(Mod(1.3, 1), .3) and the result will be True.

# This is the 6th commit message:

Fixing conventions of LaTeX printing for rf/ff

# This is the 7th commit message:

Core: inequalities with NaN raise TypeError

NaN is indeterminant; for example the result of `0/0` could be
many values in the sense of limits (`5*x/x`, `x/x`, as `x` to zero).
By this interpretation, `3 < NaN` is not an valid comparison and
we raise the error.

Note this is different from the usual IEEE 754 Floating Point
system which regards all inequalities with NaN as false, using
the logic "3 is not determinably less than NaN".

Some tests in core/tests/test_numbers also updated.

This is discussed in issue #7773 and in pr #7776.

# This is the 8th commit message:

NaN: docstring update: doesn't exactly "represent" floating point nan

With changes to inequality behaviour, NaN behaves differently than
floating point nan.  We say "loosely related" instead.

# This is the 9th commit message:

meijerint: workaround NaN's in _check_antecedents

Inequality checks with NaN's used to be false.  Now they raise
TypeError.  This causes failures in meijerint.py's
`_check_antecedents`.  I've kludged it to get the previous behaviour
but its possible those NaN's are a symptom of a real issue.  Left
some comments and FIXME's in there for this.  Also labelled several
of the tests that hit the issue.

# This is the 10th commit message:

minor modification to meijerint

c15 might contain zoo*expr which will be either NaN or zoo -- in either
case a comparison to 0 cannot be made so this is recast as False, too.

# This is the 11th commit message:

watch for inequality TypeError in valid

# This is the 12th commit message:

fix #8984 - use \left | instead of \lvert in latex(Abs)

# This is the 13th commit message:

raise TypeError for inequalities involving zoo

and Except TypeError: invalid comparison of zoo in _simplifyconds

# This is the 14th commit message:

minor refactoring in meijerint

# This is the 15th commit message:

Fixed the sympify bug caused in python3 and updated trace test

Code quality: Deleted extra lines

Update Trace Test: added an .arg test

Restore MatMul Trace evaluation, update trace tests

MatMul, Trace: various tests and XFAIL tests

Rebasing

# This is the 16th commit message:

Removed C from integrals

# This is the 2nd commit message:

In parse_expr with implicit multiplication, check if parentheses match.

# This is the 2nd commit message:

Add tests for Extend Catalan number to negative numbers

# This is the 3rd commit message:

Removed C from series

# This is the 4th commit message:

Removed C from simplify

# This is the 5th commit message:

Removed C from solvers where the imports would not create cycles

lokeshh committed:
Switch to Add(trace) instead of trace(MatAdd)

Inorder for trace to be distributed all over the expression
to have flexibility in evaluation, distributing MatAdd over
trace wasn't helpful.

lokeshh committed:
Merge branch '9028_trace_matadd'

lokeshh commented:
Very impressive! after applying #9009 everything seems good now. All examples I can think of are passing like:
```python
>>> X = Matrix([2])
>>> Y = MatrixSymbol('Y', 1, 1)
>>> trace(MatAdd(X, Y, -3*Y))
2 - 2*Trace(Y)
```
> Well perhaps we don't need to fix everything all at once ;-) I like the idea of adding an XFAIL test at least for your first X + Y - Y example. You could file an issue, then put the issue number as a comment in the test. #9009 is merged so this PR could also remove the XFAIL decorator from my test in tests/test_trace

I am on it.

> Also: take a look at the `srepr` for your second example.

Where do I look for `srepr`?

cbm755 commented:
> Where do I look for `srepr`?

Its a command that shows you the string representation.  E.g.,

````
U = MatixSymbol('U', 2, 2)
srepr(U)
````

smichr committed:
minor modification to meijerint

c15 might contain zoo*expr which will be either NaN or zoo -- in either
case a comparison to 0 cannot be made so this is recast as False, too.

leosartaj committed:
fix #8984 - use \left | instead of \lvert in latex(Abs)

smichr committed:
watch for inequality TypeError in valid

aktech committed:
raise TypeError for inequalities involving zoo

and Except TypeError: invalid comparison of zoo in _simplifyconds

lokeshh commented:
```python
@XFAIL
def test_Trace_MatAdd_doit():
    # FIXME: Issue #9028.
    X = Matrix([[1, 2], [3, 4]])
    q = MatAdd(X, 2*X)
    assert Trace(q).doit(deep=False).arg == q
```

I couldn't pass this test since Trace gets distributed when applied on MatAdd.

There's another problem which I am trying to resolve and that is:
```python
trace(Y) + trace(-Y)
```
is returning
```python
-Trace(Y) + Trace(Y)
```
BTW `srepr` is proving to be very useful. Thanks for that!

lokeshh committed:
Trace: Improved expression evaluation involving Trace

Previouslly, expression involding trace weren't giving the right results.
For example, trace(Y) + trace(-Y) wasn't giving the desired result
that is zero because trace(-Y) was equivalent to
"Mul(Integer(-1), Trace(MatMul(MatrixSymbol('Y', Integer(1), Integer(1)))))"
The presence of MatMul was not required and was the reason things weren't going
as they should.

It has been rectified by modifying the MatMul._eval_trace function.
The code is much cleaner, easy to understand and giving correct results now.
Now trace(-Y) is equivalent to
"Mul(Integer(-1), Trace(MatrixSymbol('Y', Integer(1), Integer(1))))"
which is both correct and upon addition with trace(Y) is quivalent to zero
like it should be.

lokeshh committed:
Trace: Improved expression evaluation involving Trace

Previouslly, expression involding trace weren't giving the right results.
For example, trace(Y) + trace(-Y) wasn't giving the desired result
that is zero because trace(-Y) was equivalent to
"Mul(Integer(-1), Trace(MatMul(MatrixSymbol('Y', Integer(1), Integer(1)))))"
The presence of MatMul was not required and was the reason things weren't going
as they should.

It has been rectified by modifying the MatMul._eval_trace function.
The code is much cleaner, easy to understand and giving correct results now.
Now trace(-Y) is equivalent to
"Mul(Integer(-1), Trace(MatrixSymbol('Y', Integer(1), Integer(1))))"
which is both correct and upon addition with trace(Y) is quivalent to zero
like it should be.

cbm755 commented:
1.  I will think about my test...

2.  I think that (uppercase) `Trace(Y) + Trace(-Y)` is supposed to stay unevaluated: but I think it could go to `Trace(Y) - Trace(Y)`.  (lowercase trace is a different matter).  Both should be tested.

3.  Suppose `X` is a concrete ImmutableMatrix (like the 2x2 in my test).  Then `Trace(X) + Trace(-X)` should also not evaluate.  With `.doit()` it should go to zero.  I'm not sure about `.doit(deep=False)`, probably should canonicalize but not evaluate (whatever that means in this context!)   Actually `Trace(X) + Trace(-X)` looks like another bug to me!

4.  Perhaps we are trying to do too much here, too quickly.  Are we mixing up two different issues?  Possible to just fix MatAdd?  Or are these too interconnected?

cbm755 commented on sympy/matrices/expressions/matadd.py:
```diff
@@ -51,7 +51,7 @@ def _eval_adjoint(self):
     def _eval_trace(self):
         from trace import Trace
-        return MatAdd(*[Trace(arg) for arg in self.args]).doit()
+        return Add(*[Trace(arg).doit() for arg in self.args])
```
I think I stand by my XFAIL test.

Perhaps we should drop this `.doit()`?  But will something else fail?

cbm755 replied:
Wait, maybe I stand by it not evaluating... but test might be wrong in current form.  Will think about it more...

lokeshh replied:
I am dropping the `.doit()`. It's better this way.

cbm755 replied:
Will lowercase trace() function properly without it?

lokeshh replied:
Yep

lokeshh commented:
Well, my main focus was to write tests for MatAdd but I got confused what tests should I write. Then I looked at other Classes like Adjoint, Transpose to take inspiration but came to know Trace is different from them in lot of aspects. I wished for a uniformity among them all at how they evaluate and strongly felt that Trace should be changed. For example,

```python
>>> adjoint(Y) + adjoint(-Y)
0
```
while Trace works as follows
```python
>>> trace(Y) + trace(-Y)
-Trace(Y) + Trace(Y)
```

I think I got too carried away with them and lost the main focus which was to rectify MatAdd. I think we have to refrain from writing any more tests unless we make sure what are we looking for. So, I guess I will leave it like that for now and do things one at a time. 

What do you say?

cbm755 commented:
Perhaps we can file bugs for these other issues (and also search for existing issues).   (I think I agree in principle with the Trace/trace changes, or at least with your examples!).

cbm755 commented:
Maybe this the test I want?  One must be very careful changing tests...  ImmutableMatrix is b/c of #9043.

````
def test_Trace_MatAdd_doit():
    X = ImmutableMatrix([[1, 2], [3, 4]])
    q = MatAdd(X, 2*X)
    assert Trace(q).doit(deep=False) == Trace(X) + Trace(2*X)
````

lokeshh commented:
Cool! I can pass this test.

lokeshh committed:
Update Matrice Test

lokeshh committed:
Update Matrice Test

smichr committed:
minor refactoring in meijerint

cbm755 commented on sympy/matrices/expressions/tests/test_trace.py:
```diff
@@ -53,12 +53,10 @@ def test_Trace_doit():
+    X = ImmutableMatrix([[1, 2], [3, 4]])
     q = MatAdd(X, 2*X)
-    assert Trace(q).doit(deep=False).arg == q
+    assert Trace(q).doit(deep=False) == Trace(X) + Trace(2*X)
```
might as well test that `Trace(q).doit()` gives the correct answer here too.

lokeshh replied:
Done!

cbm755 commented:
In general, you shouldn't push without checking the local test works: just do `test('sympy/matrices/expressions/tests/test_trace.py')`.  (of course, some other unexpected test can always fail).

(you need to add ImmutableMatrix to the imports at the top.)

lokeshh commented:
> In general, you shouldn't push without checking the local test works: just do test('sympy/matrices/expressions/tests/test_trace.py'). (of course, some other unexpected test can always fail).

Oops, I forgot to run the tests! it won't happen again.
BTW where am I supposed to run `test(sympy/.../test_trace.py)`?

> (you need to add ImmutableMatrix to the imports at the top.)

Thanks for mentioning that out :+1:

lokeshh committed:
Fix the trace test

Forgot to import the ImmutableMatrix module before using it
was the cause tests were passing.
It's been fixed.
Futher Trace doit test has been introduced.

lokeshh committed:
Fix the trace test

Forgot to import the ImmutableMatrix module before using it
was the cause tests were passing.
It's been fixed.
Futher Trace doit test has been introduced.

lokeshh commented:
So what are we left with now? These are I can think of:

a). Deciding what should `Trace(Y) + Trace(-Y)` and `trace(Y) + trace(-Y)` evaluate to?

b). What should `.doit(deep = False)` supposed to do? 
```python
>>> Adjoint(MatAdd(X, 2*X)).doit(deep = False)
```
gives
```python
3*Adjoint(Y)
```
while
```python
>>> Trace(MatAdd(X, 2*X)).doit(deep = False)
```

gives

```python
Trace(2*Y) + Trace(Y)
```
(I am with changing `Adjoint` and other functions to match the `Trace.doit(deep=False)` situation which seems more intuitive. What do you say on this?)


c). #9043

(What else? Will add more to this list as they come up)

lokeshh committed:
Merge branch 'master' of git://github.com/sympy/sympy into 9028_trace_matadd

lokeshh commented:
I have no idea why the test is failing. It's working fine in my system.

cbm755 commented:
failing test: something different in python 3 apparently, not sure either.  Install on Python 3 and play with it?

I don't see a test yet for this: https://github.com/sympy/sympy/pull/9031#issuecomment-74974127

If I were working on this, I would go read up on what deep=False is supposed to do in other cases (away from MatrixExpr).

lokeshh commented:
Yep, its failing with Python3 as your suggested. `trace.Trace` object is passed to sympify instead of `sympy.matrices.expressions.trace.Trace`. Any idea why?

Where can I read about function of deep = False?

cbm755 commented:
1.  No idea :(

2.  maybe grep source code for `deep`?

lokeshh commented:
Much of the code still uses old-style classes where as support of old-style classes has been removed from Python3. I don't know if that's what is causing the bug but shouldn't the whole code need to be revisited to replace the old-style classes with new style classes?

lokeshh commented:
Hahaha... you won't believe what happened!!!

I figured out the bug. It because of this:
`.`

Phew... I read whole bunch of pages about differences in Python2 and Python3 and in the end the problem was just a dot.

lokeshh committed:
Fixed the sympify bug caused in python3 and updated trace test

Code quality: Deleted extra lines

Update Trace Test: added an .arg test

Restore MatMul Trace evaluation, update trace tests

MatMul, Trace: various tests and XFAIL tests

Rebasing

lokeshh committed:
Fixed the sympify bug caused in python3 and updated trace test

lokeshh committed:
Code quality: Deleted extra lines

lokeshh committed:
Update Trace Test: added an .arg test

cbm755 commented:
Heh :)

I think you should drop this MatMul change: its incorrect as is, b/c `tr(A*B)` is not `tr(A)*tr(B)` in general.  We should probably add a test to protect this property.

I've thought about `Trace.doit()`:

1.  I the only way I can see to make deep=False work is to add a deep vs shallow kwarg to `._eval_trace` or add some kind of `._shallow_trace`.  I do not think we should do that here in this PR.  Maybe we just drop `deep=False` support altogether in `Trace.doit()` but anyway not for this PR.
2.  So my feeling is we leave an XFAIL test for Trace doit `deep=False`.
3.  We should use lowercase `trace` in your patch to `MatAdd._eval_trace()`.  I don't think `trace()` cmd can work correctly otherwise.

I'll think about some tests too.

lokeshh committed:
Restore MatMul Trace evaluation, update trace tests

lokeshh commented:
Hmm... that was a very big mistake. I should be more careful. Thanks for mentioning that out!

cbm755 committed:
MatMul, Trace: various tests and XFAIL tests

cbm755 commented:
I sent you a PR against your PR :)

lokeshh committed:
Merge pull request #1 from cbm755/9028_patches

MatMul, Trace: various tests and XFAIL tests

cbm755 commented on sympy/matrices/expressions/tests/test_trace.py:
```diff
@@ -44,21 +44,55 @@ def test_Trace():
+    q = MatAdd(X, 2*X, Y, -3*Y)
+    r = Trace(q)
+    assert r.arg == q
+    assert r.doit() == 18 - 2 * Trace(Y)
```
`2*Trace(Y)`, no spaces.

cbm755 commented:
Ok I'm +1 for this.  (you've mentioned various other potential changes: they could be filed elsewhere).

I think it will need a "rebase -i" to squash down to one or two commits.  See [dev guide](https://github.com/sympy/sympy/wiki/Development-workflow).

kevinventullo commented:
This appears to fail its own test? This test in particular:

```q = MatAdd(X, 2*X)```
```assert Trace(q).doit(deep=False).arg == q```

Even when `deep=False`, Trace will attempt to let its argument evaluate its own trace via `arg._eval_trace()`. MatAdd knows how to take its own trace so you end up getting a number if X is explicit.

cbm755 commented:
@kevinventullo isn't that test XFAIL'd?  My understanding is @lokeshh set out to solve #9028, which this does.

cbm755 commented:
@lokeshh are you able to rebase and squash these commits?

lokeshh committed:
Merge branch '9028_trace_matadd' of github.com:lokeshh/sympy into 9028_trace_matadd

Merging

lokeshh committed:
Merge branch '9028_trace_matadd' of github.com:lokeshh/sympy into 9028_trace_matadd

lokeshh commented:
Nope, I have got things messed up.

When I did `git rebase --interactive 28fd02e`, there were commits from other branches to. That was probably because I frequently kept synching with master. Anyway I squashed those commits, some of them weren't able to because of some thing I couldn't understand. I pushed them anyhow and ended up many push so many commits now.

lokeshh committed:
Merge branch '9028_trace_matadd' of github.com:lokeshh/sympy into 9028_trace_matadd

cbm755 commented:
Ok @lokeshh I did the squash and filed #9063.  (AFAICT, its still attributed to you.)

lokeshh commented:
Wow... You always have the rescue plan! I was worrying what's going to happen now. 

Thank you very much for helping me so many times.

I am looking forward to fix #9043 now.

cbm755 commented:
We fixed this (using @lokeshh's work) in #9063.
